File: src/main/golang/leetcode/editor/cn/[239]Sliding Window Maximum.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
type mqueue []int

func (t *mqueue) Len() int          { return len(*t) }
func (t *mqueue) Get(i int) (v int) { a := *t; v = a[i]; return }
func (t *mqueue) push(n int) {
	for t.Len() > 0 && n > t.Get(t.Len()-1) {
		a := *t
		*t = a[:a.Len()-1]
	}
	*t = append(*t, n)
}
func (t *mqueue) max() (v int) { a := *t; v = a[0]; return }
func (t *mqueue) pop(n int) {
	a := *t
	if a[0] == n {
		*t = a[1:]
	}
}

func maxSlidingWindow(nums []int, k int) (ans []int) {
	mq := &mqueue{}
	n := len(nums)
	for i := 0; i < n; i++ {
		mq.push(nums[i])
		if i >= k-1 {
			ans = append(ans, mq.max())
			mq.pop(nums[i-k+1])
		}
	}
	return
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[239]Sliding Window Maximum.py
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        from collections import deque
        q = deque()
        ans = []
        for i, num in enumerate(nums):
            while q and nums[q[-1]] < num:
                q.pop()
            q.append(i)
            if q[0] == i - k:
                q.popleft()
            if i >= k - 1:
                ans.append(nums[q[0]])
        return ans
# leetcode submit region end(Prohibit modification and deletion)
